src/testing/cmds/harness/test-collector/main.go

```diff
@@ -0,0 +1,182 @@
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"go.uber.org/zap"
+	"k8s.io/kubectl/pkg/util/slice"
+)
+
+type testOutput struct {
+	Time    time.Time
+	Action  string
+	Package string
+	Output  string
+}
+
+func main() {
+	log.InitPachctlLogger()
+	ctx := pctx.Background("test-collector")
+	tags := flag.String("tags", "", "Tags to run, for example k8s. Tests without this flag will not be selected.")
+	exclusiveTags := flag.Bool("exclusiveTags", true, "If true, ONLY tests with the specified tags will run. If false, "+
+		"the default behavior of 'go test' of including tagged and untagged tests is used.")
+	fileName := flag.String("file", "tests_to_run.csv", "Output file listing the packages and tests to run. Used by the runner script.")
+	pkg := flag.String("pkg", "./...", "Package to run defaults to all packages.")
+	threadPool := flag.Int("threads", 2, "Number of packages to collect tests from concurrently.")
+	flag.Parse()
+	err := run(ctx, *tags, *exclusiveTags, *fileName, *pkg, *threadPool)
+	if err != nil {
+		log.Error(ctx, "Error during tests splitting", zap.Error(err))
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+func run(ctx context.Context, tags string, exclusiveTags bool, fileName string, pkg string, threadPool int) error {
+	tagsArg := ""
+	if tags != "" {
+		tagsArg = fmt.Sprintf("-tags=%s", tags)
+	}
+	testIdsTagged, err := testNames(ctx, pkg, tagsArg)
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	var testIds map[string][]string
+	if exclusiveTags && tags != "" {
+		// set difference to get ONLY tagged tests
+		var err error
+		testIdsUntagged, err := testNames(ctx, pkg, "") // collect for set difference
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		testIds = subtractTestSet(testIdsTagged, testIdsUntagged)
+	} else {
+		testIds = testIdsTagged
+	}
+	log.Info(ctx, "tests and packages collected", zap.Any("tests", testIds))
+	err = outputToFile(fileName, testIds)
+	return err
+}
+
+// get tests that are tagged, but not in the untagged list since that is inclusive
+func subtractTestSet(testIdsTagged map[string][]string, testIdsUntagged map[string][]string) map[string][]string {
+	resultTests := map[string][]string{}
+	for pkg, testsTagged := range testIdsTagged {
+		testsUntagged, ok := testIdsUntagged[pkg]
+		for _, testNameTagged := range testsTagged {
+			if !ok || !slice.ContainsString(testsUntagged, testNameTagged, func(s string) string { return s }) {
+				if _, ok := resultTests[pkg]; !ok {
+					resultTests[pkg] = []string{testNameTagged}
+				} else {
+					resultTests[pkg] = append(resultTests[pkg], testNameTagged)
+				}
+			}
+		}
+
+	}
+	return resultTests
+}
+
+func testNames(ctx context.Context, pkg string, addtlCmdArgs ...string) (map[string][]string, error) {
```

Comment:
I think you probably want a unit test for all the logic that this script does.  I'd recommend taking an io.Reader as an argument to this function; in production, pass cmd.Stdout, in tests, pass a `strings.Reader` or `bytes.Buffer`.

Reply:
Done, I was on the fence about testing set subtraction logic as well so I added that too.